File: src/jimn/point.py
"""
points (2d or 3d) or vectors.
"""
from math import atan2, sqrt, cos, sin, pi
from jimn.bounding_box import BoundingBox
from jimn.utils.precision import is_almost
from jimn.utils.tour import tour


class Point:
    """
    a point is defined as a vector of any given dimension.

    for example:

    - create a point at x=2, y=5:

    my_point = Point([2, 5])

    - find distance between two points:

    distance = point1.distance_to(point2)

    - translate point by (2, 3):

    point += Point([2, 3])
    """
    def __init__(self, coordinates):
        """
        build new point using an array of coordinates.
        """
        self.coordinates = coordinates

    def distance_to(self, other):
        """
        euclidean distance between two points.
        """
        # this is not so elegant but fastest version
        # we need it to be fast since it is called many times
        total = 0
        for c_1, c_2 in zip(self.coordinates, other.coordinates):
            diff = c_1 - c_2
            total += diff * diff
        return sqrt(total)

    def get_x(self):
        """return the first coordinate, assuming it exists."""
        return self.coordinates[0]

    def get_y(self):
        """return the second coordinate, assuming it exists."""
        return self.coordinates[1]

    def get_z(self):
        """return the third coordinate, assuming it exists."""
        return self.coordinates[2]

    def angle_with(self, other):
        """
        angle formed by two points and horizontal line
        passing through first point.
        angles are computed with respect to svg orientation.
        """
        coordinates_differences = [
            other.coordinates[i] - self.coordinates[i]
            for i in (1, 0)
        ]
        raw_angle = -atan2(*coordinates_differences)
        if raw_angle <= 0:
            raw_angle += 2*pi
        return raw_angle

    def get_bounding_box(self):
        """
        return min bounding box containing point.
        this method is defined on any displayable object.
        """
        return BoundingBox(self.coordinates, self.coordinates)

    def save_svg_content(self, display, color):
        """
        svg display for tycat.
        """
        svg_coordinates = display.convert_coordinates(self.coordinates)
        stroke_width = 2*display.stroke_width()
        display.write("<circle cx=\"{}\" cy=\"{}\"".format(*svg_coordinates))
        display.write(" r=\"{}\" fill=\"{}\" opacity=\"0.5\"/>\n".format(
            stroke_width, color))

    def rotate(self, angle):
        """
        rotate a 2d point around origin.
        """
        cosinus = cos(-angle)
        sinus = sin(-angle)
        return Point([
            self.scalar_product(Point([cosinus, -sinus])),
            self.scalar_product(Point([sinus, cosinus]))
        ])

    def rotate_around(self, center, angle):
        """
        rotate a point around another one.
        """
        return center + (self-center).rotate(angle)

    def is_aligned_with(self, point2, point3):
        """
        are the three given points approximately aligned ?
        """
        (x_1, y_1), (x_2, y_2), (x_3, y_3) = [
            p.coordinates for p in (self, point2, point3)
        ]
        determinant = x_1*y_2 + y_1*x_3 + x_2*y_3 \
            - (y_2*x_3 + y_1*x_2 + x_1*y_3)
        return abs(determinant) < 10**-5  # TODO: why 5 ???

    def is_almost(self, other):
        """
        return true if self and other are almost the same points
        (see jimn.utils.precision)
        """
        for c_1, c_2 in zip(self.coordinates, other.coordinates):
            if not is_almost(c_1, c_2):
                return False
        return True

    def projection(self, dimension):
        """
        return the projection of the point on the first "dimension" coordinates.
        precondition : point has at least "dimension" coordinates.
        """
        return Point(self.coordinates[0:dimension])

    def scalar_product(self, other):
        """
        scalar (dot) product between two vectors
        """
        return sum([c1*c2
                    for c1, c2 in zip(self.coordinates, other.coordinates)])

    def cross_product(self, other):
        """
        cross product between 2 2d vectors.
        """
        return self.perpendicular_vector().scalar_product(other)

    def nearest_points(self, other):
        """
        dummy procedure used in graphs.
        re-defined for other more complex classes.
        """
        return (self, other)

    def perpendicular_vector(self):
        """
        return a vector perpendicular to given one.
        """
        return Point([-self.coordinates[1], self.coordinates[0]])

    def __add__(self, other):
        """
        addition operator. (useful for translations)
        """
        return Point([i + j
                      for i, j in zip(self.coordinates, other.coordinates)])

    def __sub__(self, other):
        """
        substraction operator. (useful for translations)
        """
        return Point([i - j
                      for i, j in zip(self.coordinates, other.coordinates)])

    def __mul__(self, factor):
        """
        multiplication by scalar operator. (useful for scaling)
        """
        return Point([c*factor for c in self.coordinates])

    def __truediv__(self, factor):
        """
        division by scalar operator. (useful for scaling)
        """
        return Point([c/factor for c in self.coordinates])

    def __eq__(self, other):
        """
        strict equality operator
        """
        return self.coordinates == other.coordinates

    def __hash__(self):
        return hash(tuple(self.coordinates))

    def __lt__(self, other):
        """
        Returns whether or not a < b in lexicographical order.
        Assumes a and b have the same dimension.
        """
        return self.coordinates < other.coordinates

    def __le__(self, other):
        return self.coordinates <= other.coordinates

    def __ge__(self, other):
        return self.coordinates >= other.coordinates

    def __str__(self):
        """
        print code generating the point.
        """
        return "Point([" + ', '.join([str(c) for c in self.coordinates]) + "])"


def __tour():
    """
    give a few examples.
    """
    description = "we provide a 'Point' class encoding points or vectors."
    example = """
from jimn.point import Point
from jimn.displayable import tycat
point1 = Point([0, 0])
point2 = Point([2, 3])
tycat(point1, point2)
print(\"distance between points:\", point1.distance_to(point2))
vector = Point([1, 1])
translations = (point1 + vector, point1 - vector)
tycat(point1, point2, translations)
    """

    tour("jimn.point", description, example)


if __name__ == "__main__":
    __tour()

File: rust/jimn/src/point.rs
//! Points on the plane.
//!
//! Provides a `Point` structure for storing 2d points.
//! Points can also serve as vectors: for example point2-point1 is a point
//! which coordinates encode the direction vector of segment(point1,point2).
use std::cmp::Ordering;
use std::ops::{Add, Div, Mul, Sub};
use std::fmt;

use quadrant::{Quadrant, Shape};
use utils::precision::is_almost;
use utils::float::hash_float;
use std::hash::{Hash, Hasher};


#[derive(Copy, Clone, Debug, PartialEq)]
/// 2D point structure.
/// Note that Y is defined first so that lexicographical comparisons will
/// start with y.
pub struct Point {
    /// X coordinate.
    pub x: f64,
    /// Y coordinate.
    pub y: f64,
}

#[cfg_attr(feature = "cargo-clippy", allow(derive_hash_xor_eq))]
impl Hash for Point {
    fn hash<H: Hasher>(&self, state: &mut H) {
        hash_float(&self.x, state);
        hash_float(&self.y, state);
    }
}


impl Eq for Point {}
impl Ord for Point {
    fn cmp(&self, other: &Self) -> Ordering {
        self.y
            .partial_cmp(&other.y)
            .unwrap()
            .then(self.x.partial_cmp(&other.x).unwrap())
    }
}

impl PartialOrd for Point {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Default for Point {
    fn default() -> Self {
        Point::new(0.0, 0.0)
    }
}

impl fmt::Display for Point {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{},{}", self.x, self.y)
    }
}

impl Point {
    /// Returns a new Point from given coordinates.
    pub fn new(x: f64, y: f64) -> Point {
        Point { x: x, y: y }
    }

    /// Returns if the three given points are approximately aligned.
    pub fn is_aligned_with(&self, p2: &Point, p3: &Point) -> bool {
        let determinant = self.x * p2.y + self.y * p3.x + p2.x * p3.y
            - (p2.y * p3.x + self.y * p2.x + self.x * p3.y);
        determinant.abs() < 10.0_f64.powi(-5) //TODO: why 5 ?
    }

    /// Returns distance between given points.
    ///
    /// # Example
    /// ```
    /// use jimn::point::Point;
    /// use jimn::utils::precision::is_almost;
    /// let distance = Point::new(0.0, 0.0).distance_to(&Point::new(3.0, 0.0));
    /// assert!(is_almost(distance, 3.0));
    /// ```
    pub fn distance_to(self: &Point, other: &Point) -> f64 {
        let x_diff = self.x - other.x;
        let y_diff = self.y - other.y;
        let squared_distance = x_diff * x_diff + y_diff * y_diff;
        squared_distance.sqrt()
    }

    /// Look at ourselves as a vector and return a perpendicular one.
    pub fn perpendicular_vector(&self) -> Point {
        Point::new(-self.y, self.x)
    }

    /// Returns the angle needed for going towards other.
    pub fn angle_with(&self, other: &Point) -> f64 {
        let x = other.x - self.x;
        let y = other.y - self.y;
        y.atan2(x)
    }

    /// Returns cross product between vector in self and vector in other.
    ///
    /// # Example
    /// ```
    /// use jimn::point::Point;
    /// use jimn::utils::precision::is_almost;
    /// let product = Point::new(1.0, 2.0).cross_product(&Point::new(2.0, 4.0));
    /// assert!(is_almost(product, 0.0));
    /// ```
    pub fn cross_product(&self, other: &Point) -> f64 {
        (self.x * other.y) - (self.y * other.x)
    }

    /// Check if we are nearly the same as the other point.
    pub fn is_almost(&self, other: &Point) -> bool {
        is_almost(self.x, other.x) && is_almost(self.y, other.y)
    }

    /// Return our x and y coordinates as a tuple.
    pub fn coordinates(&self) -> (f64, f64) {
        (self.x, self.y)
    }
}

impl Shape for Point {
    fn get_quadrant(&self) -> Quadrant {
        Quadrant {
            min_coordinates: vec![self.x, self.y],
            max_coordinates: vec![self.x, self.y],
        }
    }
    fn svg_string(&self) -> String {
        format!("<use xlink:href=\"#c\" x=\"{}\" y=\"{}\"/>", self.x, self.y)
    }
}

//TODO: factorize all that stuff
impl<'a> Add<&'a Point> for Point {
    type Output = Point;
    fn add(self, other: &'a Point) -> Point {
        Point {
            x: self.x + other.x,
            y: self.y + other.y,
        }
    }
}

impl<'a> Add<Point> for Point {
    type Output = Point;
    fn add(self, other: Point) -> Point {
        Point {
            x: self.x + other.x,
            y: self.y + other.y,
        }
    }
}

impl<'a> Sub<&'a Point> for Point {
    type Output = Point;
    fn sub(self, other: &'a Point) -> Point {
        Point {
            x: self.x - other.x,
            y: self.y - other.y,
        }
    }
}

impl<'a> Sub<Point> for Point {
    type Output = Point;
    fn sub(self, other: Point) -> Point {
        Point {
            x: self.x - other.x,
            y: self.y - other.y,
        }
    }
}

impl<'a, 'b> Sub<Point> for &'a Point {
    type Output = Point;
    fn sub(self, other: Point) -> Point {
        Point {
            x: self.x - other.x,
            y: self.y - other.y,
        }
    }
}

impl Mul<f64> for Point {
    type Output = Point;
    fn mul(self: Point, rhs: f64) -> Point {
        Point {
            x: self.x * rhs,
            y: self.y * rhs,
        }
    }
}

impl Div<f64> for Point {
    type Output = Point;
    fn div(self: Point, rhs: f64) -> Point {
        Point {
            x: self.x / rhs,
            y: self.y / rhs,
        }
    }
}
